ci-operator/step-registry/azure/provision/service-principal/hypershift/azure-provision-service-principal-hypershift-commands.sh

```diff
@@ -0,0 +1,82 @@
+#!/usr/bin/env bash
+
+set -euo pipefail
+
+AZURE_AUTH_LOCATION="${CLUSTER_PROFILE_DIR}/osServicePrincipal.json"
+AZURE_AUTH_CLIENT_ID="$(<"${AZURE_AUTH_LOCATION}" jq -r .clientId)"
+AZURE_AUTH_CLIENT_SECRET="$(<"${AZURE_AUTH_LOCATION}" jq -r .clientSecret)"
+AZURE_AUTH_SUBSCRIPTION_ID="$(<"${AZURE_AUTH_LOCATION}" jq -r .subscriptionId)"
+AZURE_AUTH_TENANT_ID="$(<"${AZURE_AUTH_LOCATION}" jq -r .tenantId)"
+
+az --version
+az cloud set --name AzureCloud
+az login --service-principal -u "${AZURE_AUTH_CLIENT_ID}" -p "${AZURE_AUTH_CLIENT_SECRET}" --tenant "${AZURE_AUTH_TENANT_ID}" --output none
+
+set -x
+
+SP_NAME_PREFIX="${NAMESPACE}-${UNIQUE_HASH}"
+KV_NAME=$(<"${SHARED_DIR}/azure_keyvault_name")
+RG_NSG=$(<"${SHARED_DIR}/resourcegroup_nsg")
+RG_VNET=$(<"${SHARED_DIR}/resourcegroup_vnet")
+RG_HC=$(<"${SHARED_DIR}/resourcegroup")
+COMPONENTS="azure-disk azure-file ciro cloud-provider cncc cpo ingress capz"
+
+declare -A component_to_client_id
+declare -A component_to_cert_name
+
+for component in $COMPONENTS; do
+    name="${SP_NAME_PREFIX}-${component}"
+    scopes="/subscriptions/$AZURE_AUTH_SUBSCRIPTION_ID/resourceGroups/$RG_HC"
+    if [[ $component == ingress ]]; then
+        scopes+=" /subscriptions/$AZURE_AUTH_SUBSCRIPTION_ID/resourceGroups/$RG_VNET"
+    elif [[ $component == cloud-provider ]]; then
+        scopes+=" /subscriptions/$AZURE_AUTH_SUBSCRIPTION_ID/resourceGroups/$RG_NSG"
+    fi
+
+    client_id="$(eval "az ad sp create-for-rbac --name $name --role Contributor --scopes $scopes --create-cert --cert $name --keyvault $KV_NAME --output json --only-show-errors" | jq '.appId' | sed 's/"//g')"
+    echo "az ad app delete --id $client_id" >> "${SHARED_DIR}/remove_resources_by_cli.sh"
```

Comment:
Generally for the created sp, we saved them in `"${SHARED_DIR}/azure_sp_id"`, then call `azure-deprovision-sp-and-custom-role` in deprovision stage.

Reply:
In my case, since both the app and the SP need to be removed, it seems that azure-deprovision-sp-and-custom-role may not fully meet the requirements for this situation.

Reply:
I checked `azure-provision-service-principal-minimal-permission` step where some sp created using `az ad sp create-for-rbac` command, which is really similar with the command of `client_id` creation. I am not sure the exact difference between `az ad sp delete` and `az ad app delete`. If calling `az ad sp delete` command here, what does its behavior look like ?

Reply:
Nice, thanks for the testing.

@jinyunma Do you think we should update https://github.com/openshift/release/blob/master/ci-operator/step-registry/azure/deprovision/sp-and-custom-role/azure-deprovision-sp-and-custom-role-commands.sh#L37 to use `az ad app delete` ?

Reply:
According to [azure doc](https://learn.microsoft.com/en-us/cli/azure/microsoft-graph-migration#az-ad-sp-delete), command `az ad sp delete` does not delete corresponding application and role assignment when using az cli 2.37.0 +, I tried on local host and it is indeed the case.

> One interesting thing is if it is a script issue, and exists for a long time, there is supposed to be a lot of leftover orphan app prefixed with ci-op

Same concern here, looks like that az cli in Prow can delete associated application, but the version is 2.61.0.
```
/bin/az
WARNING: You have 2 update(s) available. Consider updating your CLI installation with 'az upgrade'
azure-cli                2.61.0.post20240516044921 *

core                     2.61.0.post20240516044921 *
telemetry                1.1.0.post20240516044921

Dependencies:
msal                              1.28.0
azure-mgmt-resource               23.1.1
```
Anyway, it should be safe to change to use `az ad app delete` in deprovision stage, so that both app and sp could be deleted, right?

Reply:
@jianlinliu 

Done. There's a conflict so I'll rebase after the rehearsals finish.

Reply:
Sure, thanks for your update. It looks good to me. You can run rebase now.

Reply:
My pleasure. I've rebased the branch.